salt/transport/rabbitmq.py

```diff
@@ -0,0 +1,1243 @@
+    we can manage a pool of REQ/REP sockets-- but for now we'll just do them in serial
+
+    TODO: this implementation was adapted from ZeroMQ. Simplify this for RMQ.
+    """
+
+    def __init__(self, opts, addr, io_loop=None):
+        """
+        Create an asynchronous message client
+
+        :param dict opts: The salt opts dictionary
+        :param str addr: The interface IP address to bind to
+        :param IOLoop io_loop: A Tornado IOLoop event scheduler [tornado.ioloop.IOLoop]
+        """
+        self.opts = opts
+        self.addr = addr
+        self._rmq_connection_wrapper = RMQBlockingConnectionWrapper(opts)
+
+        if io_loop is None:
+            self.io_loop = salt.ext.tornado.ioloop.IOLoop.current()
+        else:
+            self.io_loop = io_loop
+
+        self.serial = salt.payload.Serial(self.opts)
+
+        self.send_queue = []
+        # mapping of message -> future
+        self.send_future_map = {}
+
+        self.send_timeout_map = {}  # message -> timeout
+        self._closing = False
+
+    # TODO: timeout all in-flight sessions, or error
+    def close(self):
+        try:
+            if self._closing:
+                return
+        except AttributeError:
+            # We must have been called from __del__
+            # The python interpreter has nuked most attributes already
+            return
+        else:
+            self._closing = True
+
+    # pylint: disable=W1701
+    def __del__(self):
+        self.close()
+
+    # pylint: enable=W1701
+
+    @salt.ext.tornado.gen.coroutine
+    def _internal_send_recv(self):
+        while len(self.send_queue) > 0:
+            message = self.send_queue[0]
+            future = self.send_future_map.get(message, None)
+            if future is None:
+                # Timedout
+                del self.send_queue[0]
+                continue
+
+            # send
+            def mark_future(msg):
+                if not future.done():
+                    data = self.serial.loads(msg)
+                    future.set_result(data)
+
+            # send message
+            message_correlation_id = uuid4().hex # TODO: optimize this to use a combination of hostname/port
+            self._rmq_connection_wrapper.publish(message, reply_queue_name=message_correlation_id)
+            self._rmq_connection_wrapper.consume_reply(mark_future, reply_queue_name=message_correlation_id)
+
+            try:
+                ret = yield future
+            except Exception as err:  # pylint: disable=broad-except
+                del self.send_queue[0]
+                continue
+            del self.send_queue[0]
+            self.send_future_map.pop(message, None)
+            self.remove_message_timeout(message)
+
+    def remove_message_timeout(self, message):
+        if message not in self.send_timeout_map:
+            return
+        timeout = self.send_timeout_map.pop(message, None)
```

Comment:
**suggestion** with the fallback, this if statement is unnecessary. The one just after this is already guarding for None, so... that's really all that's necessary.